src/metaprob/autodiff.cljc

```diff
@@ -0,0 +1,247 @@
+(ns metaprob.autodiff
+  (:refer-clojure :exclude [+ * - / == < > <= >= zero? number? pos? neg?]))
+
+;; Dual numbers represent computations dependent on a value with respect to which we are differentiating.
```

Comment:
Looked up "[dual number](https://en.wikipedia.org/wiki/Dual_number)" -- interesting!  If we call these records "dual numbers," then I'd probably suggest renaming `:value` and `:derivative` to something like `:macro` and `:micro` (where `:micro` is the infinitesimal part).

FWIW, I think of these "dual numbers" as being "tangent vectors" - i.e., elements of the [tangent bundle](https://en.wikipedia.org/wiki/Tangent_bundle).  In that setup, `:value` is the base point `x` and `:derivative` is the direction `y \in T_x M`.

Reply:
My thought is that the easiest way to understand what's going on in forward-mode differentiation, for someone who knows high-school calculus, is that in order to differentiate some computation with respect to some value `x`, all numerical values `v` that are functions of `x` are replaced with pairs of values `(v(x), dv/dx)`. As further operations are applied, both are updated in tandem.  At the end, if the return value is a function of `x`, it will also be such a pair, and we can read off its derivative with respect to `x` by looking at the second value.

To me, it is natural to think of the first element of the pair, `v(x)`, as the "value," and the second as the "derivative." But I do see that to others, thinking of these pairs as containing a "macro" and "micro" part might make more sense. Perhaps @zane can also weigh in.

Reply:
I agree that `:macro` and `:micro` are more confusing than `:value` and `:derivative`, but I think those field names are not congruent with the container name "dual number."  I think straight-up calling it something like `value-with-derivative` or similar would be easier to understand especially for people who (like me) had never heard of dual numbers.

Reply:
That makes sense to me! I wanted to give people a breadcrumb to search Wikipedia (if they wanted to read more about this technique), but we could also do that in a comment.